File: backend/app/Services/LiveEventService.php
<?php

namespace App\Services;

use App\Models\LiveEvent;

class LiveEventService
{
    public function record(
        string $eventType,
        string $message,
        ?int $organizationId = null,
        array $context = [],
    ): LiveEvent {
        return LiveEvent::create([
            'organization_id' => $organizationId,
            'event_type' => $eventType,
            'message' => $message,
            'payload' => $context['payload'] ?? null,
            'camera_id' => $context['camera_id'] ?? null,
            'employee_id' => $context['employee_id'] ?? null,
            'visitor_id' => $context['visitor_id'] ?? null,
            'access_point_id' => $context['access_point_id'] ?? null,
            'occurred_at' => $context['occurred_at'] ?? now(),
        ]);
    }
}

File: backend/app/services/live_event_service.py
from __future__ import annotations

import json
from datetime import datetime, timezone

from sqlalchemy.ext.asyncio import AsyncSession

from app.models.live_event import LiveEvent
from app.realtime.hub import emit


async def create_live_event(
    db: AsyncSession,
    organization_id: int | None,
    event_type: str,
    message: str,
    payload: dict | None = None,
    camera_id: int | None = None,
    employee_id: int | None = None,
    visitor_id: int | None = None,
    access_point_id: int | None = None,
) -> None:
    event = LiveEvent(
        organization_id=organization_id,
        event_type=event_type,
        message=message,
        payload=json.dumps(payload) if payload else None,
        camera_id=camera_id,
        employee_id=employee_id,
        visitor_id=visitor_id,
        access_point_id=access_point_id,
        occurred_at=datetime.now(timezone.utc),
    )
    db.add(event)
    await db.flush()

    await emit(
        organization_id,
        event_type,
        {
            "message": message,
            "payload": payload,
            "camera_id": camera_id,
            "employee_id": employee_id,
            "visitor_id": visitor_id,
            "access_point_id": access_point_id,
            "occurred_at": event.occurred_at.isoformat() if event.occurred_at else None,
        },
    )
